server/src/main/java/org/elasticsearch/persistent/AllocatedPersistentTask.java

```diff
@@ -99,30 +100,49 @@ boolean markAsCancelled() {
     }
 
     public void markAsCompleted() {
-        completeAndNotifyIfNeeded(null);
+        completeAndNotifyIfNeeded(null, null);
     }
 
     public void markAsFailed(Exception e) {
         if (CancelTasksRequest.DEFAULT_REASON.equals(getReasonCancelled())) {
-            completeAndNotifyIfNeeded(null);
+            completeAndNotifyIfNeeded(null, null);
         } else {
-            completeAndNotifyIfNeeded(e);
+            completeAndNotifyIfNeeded(e, null);
         }
     }
 
-    private void completeAndNotifyIfNeeded(@Nullable Exception failure) {
+    /**
+     * Indicates that this persistent task is no longer going to run on the local node.
+     * This will cause the local task to be terminated, and the associated persistent
+     * task to be reassigned by the master node. The persistent task <em>may</em> be
+     * reassigned to the same node unless separate measures have been taken to prevent
+     * this. The task should complete any graceful shutdown actions before calling this
+     * method.
+     * @param localAbortReason Reason for the task being aborted on this node. This
```

Comment:
I am not sure about this.  Letting the individual callers provide their own message means it can be more meaningful to the type of task.  For example, the ML ones could say something like "job relocating after gracefully persisting state due to impending node shutdown".  I am also wondering how many other teams will use this functionality.  I imagine some persistent tasks can run until the instant the node shuts down without needing to redo lots of work after relocating.  So those would never benefit from being aborted locally.  ML might be the special case due to the way persistent tasks map to completely separate processes that do non-trivial work.

So I didn't do this in the most recent round of edits.  It's not hard to add in a future PR if we determine that there's a benefit to lots of types of persistent task being proactively moved to different nodes during node shutdown and all having the same reason.

Reply:
OK.